### PR TITLE
[Merged by Bors] - feat(Order/KrullDimension): add some basic lemmas about krull dimension of a preorder

### PR DESCRIPTION
**Pull request by jjaassoonn**
- krull dimension, height, coheight interacting with functions
- order dual

These will be specialized to ring theoretic krull dimension later

---

- [x] depends on:#11730

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/KrullDimension.lean [hunk 1]
```diff
@@ -26,7 +26,20 @@ to product of varieties so that `-∞` being the Krull dimension of empty variet
...
 variable (α : Type*) [Preorder α]
+variable {β : Type*} (r : Rel β β)
+
+/--
+Krull dimension of a set `α` with a binary relation `r` is the supremum of the rightmost index of
+all relation series of `r`. If there is no relation series `a₀, a₁, ..., aₙ` in `α`, then its Krull
+dimension is defined to be negative infinity; if the length of all relation series `a₀, a₁, ..., aₙ`
+is unbounded, its Krull dimension is defined to be positive infinity. (Not sure if this definition
+is useful.)
+-/
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:40]:
```suggestion
/-- The **Krull dimension** of a set `α` with a binary relation `r` is the supremum of the rightmost
index of all relation series of `r`.

If there is no relation series `a₀, a₁, ..., aₙ` in `α`, then its Krull dimension is defined to be
negative infinity; if the length of all relation series `a₀, a₁, ..., aₙ` is unbounded, its Krull
dimension is defined to be positive infinity. (Not sure if this definition is useful.) -/
```

File: Mathlib/Order/RelSeries.lean
Comment (YaelDillies) [inline Mathlib/Order/RelSeries.lean:1]: Can you make those changes a separate PR?

**Review by YaelDillies — COMMENTED** [review]
What's your use case for `KrullDimOfRel`?

**Comment by jjaassoonn** [comment]
> What's your use case for `KrullDimOfRel`?

I have no other use of `KrullDimOfRel` other than specializing it into `KrullDim`. But I thought I could add it for the sake of generality.

**Comment by YaelDillies** [comment]
Then let's turn it into a note instead. I doubt people will ever need it.

**Comment by jjaassoonn** [comment]
> No sorry I meant "let's remove the definition and add a note about how we could have it but don't because nobody needs it".
Thanks for the clarification and sorry for the misunderstanding. Only after I removed `krullDimOfRel`, I realized that it might be useful afterall, because a composition series is a relation series of `\covby`, anyway, If it turns out to be useful, I will open another PR and revert the change I guess.

File: Mathlib/Order/KrullDimension.lean [hunk 2]
```diff
@@ -24,19 +24,26 @@ Krull dimensions are defined to take value in `WithBot (WithTop ℕ)` so that `(
...
-Krull dimension of a preorder `α` is the supremum of the rightmost index of all relation
+The **Krull dimension** of a preorder `α` is the supremum of the rightmost index of all relation
 series of `α` order by `<`. If there is no series `a₀ < a₁ < ... < aₙ` in `α`, then its Krull
 dimension is defined to be negative infinity; if the length of all series `a₀ < a₁ < ... < aₙ` is
 unbounded, its Krull dimension is defined to be positive infinity.
 -/
 noncomputable def krullDim : WithBot (WithTop ℕ) :=
   ⨆ (p : LTSeries α), p.length
 
+
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:46]:
```suggestion
```

File: Mathlib/Order/KrullDimension.lean [hunk 3]
```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
 Coheight of an element `a` of a pre-ordered set `α` is the Krull dimension of the subset `[a, +∞)`
 -/
 noncomputable def coheight (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Ici a)
+
+end definitions
+
+namespace krullDim
+
+variable {α β : Type _}
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:61]:
```suggestion
variable {α β : Type*}
```

File: Mathlib/Order/KrullDimension.lean [hunk 4]
```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
 Coheight of an element `a` of a pre-ordered set `α` is the Krull dimension of the subset `[a, +∞)`
 -/
 noncomputable def coheight (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Ici a)
+
+end definitions
+
+namespace krullDim
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:59]: I would rather not use a namespace here. `krullDim` is not a structure.

File: Mathlib/Order/KrullDimension.lean [hunk 5]
```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+
+variable {α β : Type _}
+
+variable [Preorder α] [Preorder β]
+
+lemma nonneg_of_nonempty [Nonempty α] : 0 ≤ krullDim α :=
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:65]: Instead I would prefix the lemmas with `krullDim_` (or infixed, as appropriate)
```suggestion
lemma krullDim_nonneg_of_nonempty [Nonempty α] : 0 ≤ krullDim α :=
```

File: Mathlib/Order/KrullDimension.lean [hunk 6]
```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+  le_sSup ⟨⟨0, fun _ ↦ @Nonempty.some α inferInstance, fun f ↦ f.elim0⟩, rfl⟩
+
+lemma eq_bot_of_isEmpty [IsEmpty α] : krullDim α = ⊥ := WithBot.ciSup_empty _
+
+lemma eq_top_of_infiniteDimensionalType [InfiniteDimensionalOrder α] :
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:70]:
```suggestion
lemma eq_top_of_infiniteDimensionalOrder [InfiniteDimensionalOrder α] :
```

File: Mathlib/Order/KrullDimension.lean [hunk 7]
```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+| ⊥, hm => False.elim <| by
+  haveI : Inhabited α := ⟨LTSeries.withLength _ 0 0⟩
+  exact not_le_of_lt (WithBot.bot_lt_coe _ : ⊥ < (0 : WithBot (WithTop ℕ))) <| hm default
+| some ⊤, _ => le_refl _
+| some (some m), hm => by
+  refine (not_lt_of_le (hm (LTSeries.withLength _ (m + 1))) ?_).elim
+  erw [WithBot.coe_lt_coe, WithTop.coe_lt_coe]
+  simp
+
+lemma eq_len_of_finiteDimensionalType [FiniteDimensionalOrder α] :
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:82]:
```suggestion
lemma eq_length_of_finiteDimensionalType [FiniteDimensionalOrder α] :
```

File: Mathlib/Order/KrullDimension.lean [hunk 8]
```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+  refine (LTSeries.longestOf_len_unique (default : LTSeries α) fun q ↦ show _ ≤ 0 from ?_).symm
+  by_contra r
+  exact ne_of_lt (q.step ⟨0, not_le.mp r⟩) <| Subsingleton.elim _ _
+
+/-- If `f : α → β` is a strictly monotonic function and `α` is an infinite dimensional type then so
+  is `β`. -/
+lemma infiniteDimensional_of_strictMono
+    (f : α → β) (hf : StrictMono f) [InfiniteDimensionalOrder α] :
+    InfiniteDimensionalOrder β :=
+  ⟨fun n ↦ ⟨(LTSeries.withLength _ n).map f hf, LTSeries.length_withLength α n⟩⟩
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:100]: This should be earlier

File: Mathlib/Order/KrullDimension.lean [hunk 9]
```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+
+lemma le_of_strictMono (f : α → β) (hf : StrictMono f) : krullDim α ≤ krullDim β :=
+  iSup_le <| fun p ↦ le_sSup ⟨p.map f hf, rfl⟩
+
+lemma height_mono {a b : α} (h : a ≤ b) : height α a ≤ height α b :=
+  le_of_strictMono (fun x ↦ ⟨x, le_trans x.2 h⟩) <| fun _ _ h ↦ h
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:106]: This should be earlier

File: Mathlib/Order/KrullDimension.lean [hunk 10]
```diff
@@ -46,3 +53,81 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+
+lemma le_orderDual : krullDim α ≤ krullDim αᵒᵈ :=
+  iSup_le <| fun i ↦ le_sSup <| ⟨i.reverse, rfl⟩
+
+lemma orderDual_le : krullDim αᵒᵈ ≤ krullDim α :=
+  le_orderDual.trans <| le_of_strictMono
+    (OrderDual.ofDual ∘ OrderDual.ofDual) <| fun _ _ h ↦ h
+
+lemma eq_orderDual : krullDim α = krullDim αᵒᵈ :=
+  le_antisymm le_orderDual <| orderDual_le
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:131]: Inline the first two into the last one

File: Mathlib/Order/KrullDimension.lean (hunk 4 above)
Comment (jjaassoonn) [reply Mathlib/Order/KrullDimension.lean]: this is done, thanks!

File: Mathlib/Order/KrullDimension.lean (hunk 10 above)
Comment (jjaassoonn) [reply Mathlib/Order/KrullDimension.lean]: this is done, thanks!

File: Mathlib/Order/KrullDimension.lean [hunk 11]
```diff
@@ -78,55 +78,50 @@ le_antisymm le_top <| le_iSup_iff.mpr <| fun m hm ↦ match m, hm with
...
-    (OrderDual.ofDual ∘ OrderDual.ofDual) <| fun _ _ h ↦ h
-
-lemma eq_orderDual : krullDim α = krullDim αᵒᵈ :=
-  le_antisymm le_orderDual <| orderDual_le
+    (iSup_le fun i ↦ le_iSup_of_le (i ⟨i.length, lt_add_one _⟩) <|
+    le_sSup ⟨⟨_, fun m ↦ ⟨i m, i.strictMono.monotone <| show m.1 ≤ i.length by omega⟩,
+      i.step⟩, rfl⟩) <|
+    iSup_le fun a ↦ krullDim_le_of_strictMono Subtype.val fun _ _ h ↦ h
+
+lemma krullDim_eq_orderDual : krullDim α = krullDim αᵒᵈ :=
```
Comment (YaelDillies) [inline Mathlib/Order/KrullDimension.lean:123]:
```suggestion
@[simp] lemma krullDim_orderDual : krullDim αᵒᵈ = krullDim α :=
```

File: Mathlib/Order/KrullDimension.lean (hunk 8 above)
Comment (jjaassoonn) [reply Mathlib/Order/KrullDimension.lean]: I have moved this to another file.

File: Mathlib/Order/KrullDimension.lean (hunk 9 above)
Comment (jjaassoonn) [reply Mathlib/Order/KrullDimension.lean]: I moved this a bit earlier, but since it requires `krullDim_le_of_strictMono`, I can’t move it too early

**Comment by jjaassoonn** [comment]
Sorry I completely forgot about this PR.

File: Mathlib/Order/KrullDimension.lean [hunk 12]
```diff
@@ -24,12 +24,18 @@ Krull dimensions are defined to take value in `WithBot (WithTop ℕ)` so that `(
 also negative infinity. This is because we want Krull dimensions to be additive with respect
 to product of varieties so that `-∞` being the Krull dimension of empty variety is equal to
 sum of `-∞` and the Krull dimension of any other varieties.
+
+We could generalize the notion of Krull dimension to an arbitrary binary relation; many results
+in this file would generalize as well. But we don't think it would useful, so we only define
+Krull dimension of a partial order.
```
Comment (kbuzzard) [inline Mathlib/Order/KrullDimension.lean:30]: Should this say "preorder"?

File: Mathlib/Order/KrullDimension.lean [hunk 13]
```diff
@@ -46,3 +52,69 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+variable [Preorder α] [Preorder β]
+
+lemma krullDim_nonneg_of_nonempty [Nonempty α] : 0 ≤ krullDim α :=
+  le_sSup ⟨⟨0, fun _ ↦ @Nonempty.some α inferInstance, fun f ↦ f.elim0⟩, rfl⟩
+
+lemma krullDim_eq_bot_of_isEmpty [IsEmpty α] : krullDim α = ⊥ := WithBot.ciSup_empty _
+
+lemma krullDim_eq_top_of_infiniteDimensionalOrder [InfiniteDimensionalOrder α] :
+    krullDim α = ⊤ :=
+le_antisymm le_top <| le_iSup_iff.mpr <| fun m hm ↦ match m, hm with
```
Comment (kbuzzard) [inline Mathlib/Order/KrullDimension.lean:71]: Everything needs indenting +2, right?

File: Mathlib/Order/KrullDimension.lean [hunk 14]
```diff
@@ -46,3 +52,69 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+
+lemma krullDim_le_of_strictMono (f : α → β) (hf : StrictMono f) : krullDim α ≤ krullDim β :=
+  iSup_le <| fun p ↦ le_sSup ⟨p.map f hf, rfl⟩
+
+lemma height_mono {a b : α} (h : a ≤ b) : height α a ≤ height α b :=
+  krullDim_le_of_strictMono (fun x ↦ ⟨x, le_trans x.2 h⟩) <| fun _ _ ↦ id
+
+lemma krullDim_eq_length_of_finiteDimensionalOrder [FiniteDimensionalOrder α] :
+    krullDim α = (LTSeries.longestOf α).length :=
+le_antisymm
```
Comment (kbuzzard) [inline Mathlib/Order/KrullDimension.lean:89]: indent

File: Mathlib/Order/KrullDimension.lean [hunk 15]
```diff
@@ -46,3 +52,69 @@ noncomputable def height (a : α) : WithBot (WithTop ℕ) := krullDim (Set.Iic a
...
+  (iSup_le <| fun _ ↦ WithBot.coe_le_coe.mpr <| WithTop.coe_le_coe.mpr <|
+    RelSeries.length_le_length_longestOf _ _) <|
+  le_iSup (fun (i : LTSeries _) ↦ (i.length : WithBot (WithTop ℕ))) <| LTSeries.longestOf _
+
+lemma krullDim_eq_zero_of_unique [Unique α] : krullDim α = 0 :=  by
```
Comment (kbuzzard) [inline Mathlib/Order/KrullDimension.lean:94]:
```suggestion
lemma krullDim_eq_zero_of_unique [Unique α] : krullDim α = 0 := by
```

**Review by kbuzzard — COMMENTED** [review]
This looks almost ready to go, I've made some minor comments.

File: Mathlib/Order/KrullDimension.lean (hunk 12 above)
Comment (jjaassoonn) [reply Mathlib/Order/KrullDimension.lean]: Indeed, changed

File: Mathlib/Order/KrullDimension.lean (hunk 13 above)
Comment (jjaassoonn) [reply Mathlib/Order/KrullDimension.lean]: yes!

File: Mathlib/Order/KrullDimension.lean (hunk 14 above)
Comment (jjaassoonn) [reply Mathlib/Order/KrullDimension.lean]: done

**Comment by jjaassoonn** [comment]
I don't understand why CI fails

File: Mathlib/Order/KrullDimension.lean [hunk 16]
```diff
@@ -24,12 +24,18 @@ Krull dimensions are defined to take value in `WithBot (WithTop ℕ)` so that `(
 also negative infinity. This is because we want Krull dimensions to be additive with respect
 to product of varieties so that `-∞` being the Krull dimension of empty variety is equal to
 sum of `-∞` and the Krull dimension of any other varieties.
+
+We could generalize the notion of Krull dimension to an arbitrary binary relation; many results
+in this file would generalize as well. But we don't think it would useful, so we only define
```
Comment (kbuzzard) [inline Mathlib/Order/KrullDimension.lean:29]:
```suggestion
in this file would generalize as well. But we don't think it would be useful, so we only define
```

**Review by kbuzzard — APPROVED** [review]
There's a minor grammar slip in the module docstring which should be fixed, but other than that this looks fine. 

bors d+
